File: src/Simbuino.Emulator/MemoryMappedRegister.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace Simbuino.Emulator
{
	// this is a memory-mapped register class that is used mainly to provide change notification for when a register changes,
	// it's needed by devices (LCD etc) that need to know when a memory location has been written to so they can react accordingly.

	public interface IMemoryMappedRegister : Register
	{
	}

	public class MemoryMappedRegister : Register
	{
		private int Index;
		public MemoryMappedRegister(int index)
		{
			this.Index = index;
		}

		public int Value
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get { return AtmelContext.R[Index]; }
			
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set
			{
				int oldVal = AtmelContext.R[Index];
				if (oldVal != value)
					AtmelContext.R[Index] = value & 0xff;
			}
		}

		public int this[int index]
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get
			{
				return ((AtmelContext.R[Index] >> index) & 1);
			}
			
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set
			{
				if (value == 0)
					this.Value = this.Value & ~(1 << index);
				else
					this.Value = this.Value | (1 << index);
			}
		}

		public void Reset()
		{
			AtmelContext.R[Index] = 0;
		}
	}

	
	public delegate void RegisterChangedHandler(int oldVal, int newVal);
	public delegate void RegisterReadHandler(ref int val);

	public class ObservableRegister : Register
	{
		private int Index;

		public event RegisterChangedHandler OnRegisterChanged;
		public event RegisterReadHandler OnRegisterRead;

		public ObservableRegister(int index)
		{
			this.Index = index;
		}

		public int Value
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get
			{
				var val = AtmelContext.R[Index];
				if (this.OnRegisterRead != null)
					this.OnRegisterRead(ref val);
				return val;
			}

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set
			{
				int oldVal = AtmelContext.R[Index];
				AtmelContext.R[Index] = value & 0xff;
				if (this.OnRegisterChanged != null)
					this.OnRegisterChanged(oldVal, value);
			}
		}

		public int this[int index]
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get
			{
				return ((AtmelContext.R[Index] >> index) & 1);
			}

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set
			{
				if (value == 0)
					this.Value = this.Value & ~(1 << index);
				else
					this.Value = this.Value | (1 << index);
			}
		}

		public void Reset()
		{
			AtmelContext.R[Index] = 0;
		}
	}

	public class MemoryMappedWordRegister : WordRegister
	{
		private int Index;

		public MemoryMappedWordRegister(int index)
		{
			this.Index = index;
		}

		public ushort Value
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get
			{
				int lo = AtmelContext.R[this.Index];
				int hi = AtmelContext.R[this.Index + 1];
				return (ushort)(lo | (hi << 8));
			}

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set
			{
				AtmelContext.R[this.Index] = (byte)(value & 0xff);
				AtmelContext.R[this.Index + 1] = (byte)(value >> 8);
			}
		}

		public int this[int index]
		{
			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			get
			{
				return ((Value >> index) & 1);
			}

			[MethodImplAttribute(MethodImplOptions.AggressiveInlining)]
			set
			{
				if (value == 0)
					Value = (ushort)(Value & ~(1 << index));
				else
					Value = (ushort)(Value | (1 << index));
			}
		}
	}

}

File: binaries/Simbuino4Web/Scripts/Views/MemoryMappedRegister.js
$(function () {

	// this is a memory-mapped register class that is used mainly to provide change notification for when a register changes,
	// it's needed by devices (LCD etc) that need to know when a memory location has been written to so they can react accordingly.

	MemoryMappedRegister = Class.create({

		ctor: function (index) {
			this.Index = index;
		},

		get: function()
		{
			return AtmelContext.R[this.Index];
		},

		set: function(value)
		{
			var oldVal = AtmelContext.R[this.Index];
			if (oldVal != value)
				AtmelContext.R[this.Index] = value & 0xff;
		},

		get_bit: function(index)
		{
			return ((AtmelContext.R[this.Index] >> index) & 1);
		},

		set_bit: function(index, value)
		{
			if (value == 0)
				this.set(this.get() & ~(1 << index));
			else
				this.set(this.get() | (1 << index));
		},

		Reset: function()
		{
			AtmelContext.R[this.Index] = 0;
		},
	});


	ObservableRegister = Class.create({

		ctor: function(index)
		{
			this.Index = index;
			this.OnRegisterChanged = [];
			this.OnRegisterRead = [];
		},

		get: function()
		{
			var val = AtmelContext.R[this.Index];
			if (this.OnRegisterRead != null)
				for (var i=0; i<this.OnRegisterRead.length; i++)
					val = this.OnRegisterRead[i](val);
			return val;
		},

		set: function(value)
		{
			var oldVal = AtmelContext.R[this.Index];
			AtmelContext.R[this.Index] = value & 0xff;
			if (this.OnRegisterChanged != null)
				for (var i = 0; i < this.OnRegisterChanged.length; i++)
					this.OnRegisterChanged[i](oldVal, value);
		},

		get_bit: function(index)
		{
			return ((AtmelContext.R[this.Index] >> index) & 1);
		},

		set_bit: function(index, value)
		{
			if (value == 0)
				this.set(this.get() & ~(1 << index));
			else
				this.set(this.get() | (1 << index));
		},

		Reset: function()
		{
			AtmelContext.R[this.Index] = 0;
		}
	});

	MemoryMappedWordRegister = Class.create({

		ctor: function(index)
		{
			this.Index = index;
		},

		Reset: function()
		{
			this.set(0);
		},

		get: function()
		{
			var lo = AtmelContext.R[this.Index];
			var hi = AtmelContext.R[this.Index + 1];
			return (lo | (hi << 8));
		},

		set: function(value)
		{
			AtmelContext.R[this.Index] = value & 0xff;
			AtmelContext.R[this.Index + 1] = (value >> 8) & 0xff;
		},

		get_bit: function(index)
		{
			return ((this.get() >> index) & 1);
		},

		set_bit: function(index, value)
		{
			if (value == 0)
				this.set((this.get() & ~(1 << index)));
			else
				this.set((this.get() | (1 << index)));
		}
		
	});

});